eZ/Publish/Core/Limitation/ObjectStateLimitationType.php

```diff
@@ -156,20 +158,60 @@ public function evaluate(APILimitationValue $value, APIUserReference $currentUse
                     );
                 }
 
-                $objectStateIdArray[] = $defaultStateId;
+                if ($this->isStateGroupUsedForLimitation($stateGroup->id, $limitationValues)) {
+                    $objectStateIdsToVerify[] = $defaultStateId;
+                }
             }
         } else {
-            /*
-             * @var $object ContentInfo
-             */
             foreach ($stateGroups as $stateGroup) {
-                $objectStateIdArray[] = $objectStateHandler->getContentState($object->id, $stateGroup->id)->id;
+                if ($this->isStateGroupUsedForLimitation($stateGroup->id, $limitationValues)) {
+                    $objectStateIdsToVerify[] = $objectStateHandler->getContentState($object->id, $stateGroup->id)->id;
+                }
             }
         }
 
-        $intersect = array_intersect($value->limitationValues, $objectStateIdArray);
+        return $this->areObjectStatesMatchingTheLimitation($objectStateIdsToVerify, $limitationValues);
+    }
+
+    /**
+     * Checks if the State Group contains at least one State that is used by Limitation.
+     *
+     * @param int $stateGroupId
+     * @param int[] $limitationValues
+     *
+     * @return bool
+     */
+    private function isStateGroupUsedForLimitation($stateGroupId, array $limitationValues)
+    {
+        $objectStateHandler = $this->persistence->objectStateHandler();
+        $states = $objectStateHandler->loadObjectStates($stateGroupId);
+
+        foreach ($states as $state) {
+            if (in_array($state->id, $limitationValues, false)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Verifies if all the Object States are matching the Limitation Values.
+     *
+     * @param int[] $objectStateIds
+     * @param int[] $limitationValues
+     *
+     * @return bool
+     */
+    private function areObjectStatesMatchingTheLimitation(array $objectStateIds, array $limitationValues)
+    {
+        foreach ($objectStateIds as $objectStateId) {
+            if (!in_array($objectStateId, $limitationValues, false)) {
```

Comment:
Yes, it is the default, but I put it there to point out that there is no type checking on purpose - because if it is the code won't work. But I can remove it if you think that it will be more readable this way.